### PR TITLE
Allow searching PermissionType constant by Permission class name [HZ-2116]

### PR DESCRIPTION
**Pull request by kwart**
This PR adds the `PermissionConfig.PermissionType` method for searching the type by the Permission class name.

The new method is used in `SecurityContextImpl` and related tests in the EE.

Related EE PR: hazelcast/hazelcast-enterprise#6416

### CODE REVIEW

File: hazelcast/src/main/java/com/hazelcast/config/PermissionConfig.java [hunk 1]
```diff
@@ -70,124 +100,126 @@ public enum PermissionType {
         /**
          * All
```
Comment (JackPGreen) [inline hazelcast/src/main/java/com/hazelcast/config/PermissionConfig.java:101]: Not related to your change, but not exactly helpful Javadoc there.

File: hazelcast/src/main/java/com/hazelcast/config/PermissionConfig.java [hunk 2]
```diff
@@ -199,6 +231,15 @@ public static PermissionType getType(String nodeName) {
             return null;
         }
 
+        public static PermissionType getTypeByPermissionClassName(String permissionClassname) {
+            for (PermissionType type : PermissionType.values()) {
```
Comment (JackPGreen) [inline hazelcast/src/main/java/com/hazelcast/config/PermissionConfig.java:235]: Would it not be better to make a `Map<String, PermissionType>` instead?

File: hazelcast/src/main/java/com/hazelcast/config/PermissionConfig.java [hunk 3]
```diff
@@ -199,6 +231,15 @@ public static PermissionType getType(String nodeName) {
             return null;
         }
 
+        public static PermissionType getTypeByPermissionClassName(String permissionClassname) {
```
Comment (JackPGreen) [inline hazelcast/src/main/java/com/hazelcast/config/PermissionConfig.java:234]: Where is this used?

**Review by JackPGreen — CHANGES_REQUESTED** [review]

File: hazelcast/src/main/java/com/hazelcast/config/PermissionConfig.java (hunk 3 above)
Comment (kwart) [reply hazelcast/src/main/java/com/hazelcast/config/PermissionConfig.java:234]: It's used in the hazelcast-enterprise. I've updated the PR description with more details and a link to the other PR.

File: hazelcast/src/main/java/com/hazelcast/config/PermissionConfig.java (hunk 2 above)
Comment (kwart) [reply hazelcast/src/main/java/com/hazelcast/config/PermissionConfig.java:235]: This way I don't need to care about thread safety. And we already have the same approach in the `getType()` method.

Comment (JackPGreen) [reply hazelcast/src/main/java/com/hazelcast/config/PermissionConfig.java:235]: > This way I don't need to care about thread safety.

If it was a static lookup it'd be ok anyway?

> And we already have the same approach in the `getType()` method.

Good point. Maybe both should change together.

**Comment by JackPGreen** [comment]
Apologies for being picky. This looks good :)

File: hazelcast/src/main/java/com/hazelcast/config/PermissionConfig.java (hunk 1 above)
Comment (kwart) [reply hazelcast/src/main/java/com/hazelcast/config/PermissionConfig.java:101]: I'll send an update commit for the `ALL` constant.
If you meant all JavaDocs within the enum, I would leave it to another PR.

**Review by olukas — APPROVED** [review]

File: hazelcast/src/main/java/com/hazelcast/config/PermissionConfig.java (hunk 2 above)
Comment (kwart) [reply hazelcast/src/main/java/com/hazelcast/config/PermissionConfig.java]: I've simplified the code, but I didn't introduce additional data structure for the reverse lookups. IMO the enum size is small enough to be OK to do scan-through on the lookups.

**Review by JamesHazelcast — APPROVED** [review]
LGTM 👍

**Comment by kwart** [comment]
Thanks guys for all the reviews :bow:

**Pull request merged**